File: android/src/main/java/com/ztove/plugins/nativeaudio/NativeAudioSettingsPlugin.java
package com.ztove.plugins.nativeaudio;

import com.getcapacitor.JSObject;
import com.getcapacitor.Bridge;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import android.content.Context;
import android.media.AudioManager;
import android.util.Log;

import android.content.BroadcastReceiver;
import android.content.Intent;
import android.content.IntentFilter;

@CapacitorPlugin(name = "NativeAudioSettings")
public class NativeAudioSettingsPlugin extends Plugin {
    private AudioManager audioManager;
    private int lastKnownVolume = -1;

    private NativeAudioSettings implementation = new NativeAudioSettings();

    @Override
    public void load() {
        audioManager = (AudioManager) bridge.getActivity().getSystemService(Context.AUDIO_SERVICE);

        BroadcastReceiver listener = new BroadcastReceiver() {
            @Override
            public void onReceive(Context context, Intent intent) {
                if ("android.media.VOLUME_CHANGED_ACTION".equals(intent.getAction())) {
                    int notificationVolume = audioManager.getStreamVolume(AudioManager.STREAM_NOTIFICATION);
                    int maxNotificationVolume = audioManager.getStreamMaxVolume(AudioManager.STREAM_NOTIFICATION);

                    if (notificationVolume != lastKnownVolume) {
                        lastKnownVolume = notificationVolume;
                        JSObject eventData = generateNotificationVolumeChangeEvent(notificationVolume, maxNotificationVolume);
                        notifyListeners("notificationVolumeChange", eventData);
                    }
                }
            }
        };
        bridge.getActivity().registerReceiver(listener, new IntentFilter("android.media.VOLUME_CHANGED_ACTION"));
    }


    @PluginMethod
    public void echo(PluginCall call) {
        String value = call.getString("value");

        JSObject ret = new JSObject();
        ret.put("value", implementation.echo(value));
        call.resolve(ret);
    }

    @PluginMethod()
    public void getNotificationVolume(PluginCall call) {
        Context context = this.bridge.getActivity();

        AudioManager audioManager = (AudioManager) context.getSystemService(Context.AUDIO_SERVICE);
        int notificationVolume = audioManager.getStreamVolume(AudioManager.STREAM_NOTIFICATION);
        int maxNotificationVolume = audioManager.getStreamMaxVolume(AudioManager.STREAM_NOTIFICATION);
    
        JSObject ret = new JSObject();
        ret.put("notificationVolume", notificationVolume);
        ret.put("maxNotificationVolume", maxNotificationVolume);
    
        call.resolve(ret);
    }

    @PluginMethod()
    public void getMainVolume(PluginCall call) {
        Context context = this.bridge.getActivity();

        AudioManager audioManager = (AudioManager) context.getSystemService(Context.AUDIO_SERVICE);
        int mainVolume = audioManager.getStreamVolume(AudioManager.STREAM_MUSIC);
        int maxMainVolume = audioManager.getStreamMaxVolume(AudioManager.STREAM_MUSIC);

        JSObject ret = new JSObject();
        ret.put("mainVolume", mainVolume);
        ret.put("maxMainVolume", maxMainVolume);

        call.resolve(ret);
    }

    private JSObject generateNotificationVolumeChangeEvent(int volume, int maxVolume) {
        var ret = new JSObject();
        ret.put("notificationVolume", volume);
        ret.put("maxNotificationVolume", maxVolume);
        return ret;
    }
}

File: ios/Plugin/NativeAudioSettingsPlugin.swift
import Foundation
import AVFoundation
import Capacitor
import MediaPlayer

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(NativeAudioSettingsPlugin)
public class NativeAudioSettingsPlugin: CAPPlugin {
    private let implementation = NativeAudioSettings()

    @objc func echo(_ call: CAPPluginCall) {
        let value = call.getString("value") ?? ""
        call.resolve([
            "value": implementation.echo(value)
        ])
    }

      override public func load() {
        let volumeChangeNotificationName = NSNotification.Name(rawValue: "AVSystemController_SystemVolumeDidChangeNotification")
        NotificationCenter.default.addObserver(self, selector: #selector(handleVolumeChange(notification:)), name: volumeChangeNotificationName, object: nil)
    }

    private func emitNotificationVolumeChangeEvent(volume: Float, maxVolume: Float) {
        let eventData: [String: Any] = ["notificationVolume": volume, "maxNotificationVolume": maxVolume]
        notifyListeners("notificationVolumeChange", data: eventData)
    }

    @objc func getMainVolume(_ call: CAPPluginCall) {
        let audioManager = AVAudioSession.sharedInstance()
        
        let mainVolume = audioManager.outputVolume
        let maxMainVolume = getMaxVolume()

        let result: [String: Any] = ["mainVolume": mainVolume, "maxMainVolume": maxMainVolume]
        call.resolve(result)
    }

    @objc func getNotificationVolume(_ call: CAPPluginCall) {
        let audioManager = AVAudioSession.sharedInstance()
        let notificationVolume = audioManager.outputVolume
        let maxNotificationVolume = getMaxVolume()

        let result: [String: Any] = ["notificationVolume": notificationVolume, "maxNotificationVolume": maxNotificationVolume]
        call.resolve(result)
    }

    @objc func handleVolumeChange(notification: NSNotification) {
    if let userInfo = notification.userInfo,
       let changeReason = userInfo["AVSystemController_AudioVolumeChangeReasonNotificationParameter"] as? String {
        if changeReason == "ExplicitVolumeChange" {
            let audioManager = AVAudioSession.sharedInstance()
            let volume = audioManager.outputVolume
            let maxVolume = getMaxVolume()
            
            let eventData: [String: Any] = ["notificationVolume": volume, "maxNotificationVolume": maxVolume]
            notifyListeners("notificationVolumeChange", data: eventData)
            self.notifyListeners("notificationVolumeChange", data: ["testData": "Test event data"])

            }
        }
    }

    private func getMaxVolume() -> Float {
        let volumeView = MPVolumeView()
        var maxVolume: Float = 1.0

        if let view = volumeView.subviews.first as? UISlider {
            maxVolume = view.maximumValue
        }

        return maxVolume
    }
}
